pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/VarLengthValueWriter.java

```diff
@@ -0,0 +1,134 @@
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import org.apache.pinot.segment.spi.memory.CleanerUtil;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+
+/**
+ * The value writer for var-length values (STRING and BYTES).
+ *
+ * The layout of the file is as follows:
+ * <p>
+ * Header Section:
+ * <ul>
+ *   <li>
+ *     Magic bytes: Chose this to be ".vl;" to avoid conflicts with the fixed size value buffer implementations. By
+ *     having special characters, this avoids conflicts with regular bytes/strings dictionaries.
+ *   </li>
+ *   <li>
+ *     Version number: This is an integer and can be used for the evolution of the store implementation by incrementing
+ *     version for every incompatible change to the store/format.
+ *   </li>
+ *   <li>
+ *     Number of elements in the store.
+ *   </li>
+ *   <li>
+ *     The offset where the data section starts. Though the data section usually starts right after he header, having
+ *     this explicitly will let the store to be evolved freely without any assumptions about where the data section
+ *     starts.
+ *   </li>
+ * </ul>
+ * <p>
+ * Data section:
+ * <ul>
+ *   <li>
+ *     Offsets Array: Integer offsets of the start position of the byte arrays.
+ *     Example: [O(0), O(1),...O(n)] where O is the Offset function.
+ *     Length of nth element is computed as: O(n+1) - O(n). Since the last element's length can't be computed using this
+ *     formula, we store an extra offset at the end to be able to compute last element's length with the same formula,
+ *     without depending on underlying buffer's size.
+ *   </li>
+ *   <li>
+ *     All byte arrays.
+ *   </li>
+ * </ul>
+ *
+ * @see FixedByteValueReaderWriter
+ */
+public class VarLengthValueWriter implements Closeable {
+
+  /**
+   * Magic bytes used to identify the dictionary files written in variable length bytes format.
+   */
+  static final byte[] MAGIC_BYTES = ".vl;".getBytes(UTF_8);
+
+  /**
+   * Increment this version if there are any structural changes in the store format and
+   * deal with backward compatibility correctly based on old versions.
+   */
+  static final int VERSION = 1;
+
+  // Offsets of different fields in the header. Having as constants for readability.
+  static final int VERSION_OFFSET = MAGIC_BYTES.length;
+  static final int NUM_VALUES_OFFSET = VERSION_OFFSET + Integer.BYTES;
+  static final int DATA_SECTION_OFFSET_POSITION = NUM_VALUES_OFFSET + Integer.BYTES;
+  static final int HEADER_LENGTH = DATA_SECTION_OFFSET_POSITION + Integer.BYTES;
+
+  private final FileChannel _fileChannel;
+  private final ByteBuffer _offsetBuffer;
+  private final ByteBuffer _valueBuffer;
+
+  public VarLengthValueWriter(File outputFile, int numValues)
+      throws IOException {
+    _fileChannel = new RandomAccessFile(outputFile, "rw").getChannel();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code assigns a resource that might leak to a class field, causing a resource leak that slows down or crashes your system. For example, when FileChannel.map throws an exception, the *VarLengthValueWriter* object isn’t constructed so you can’t close this resource afterward.

**Fix**
Ensure that if an exception is thrown, the *\_fileChannel* resource is closed before the constructor is exited.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).